### DISCUSSION TITLE
Performance improvements to websockets

### DISCUSSION THREAD

**Discussion by tarasko** [aio-libs/aiohttp#8258] [Ideas]
Hi all,
At my current job I'm using aiohttp websockets to subscribe to updates from a 3rd party webserver. The traffic consists of a lot of 
small update messages and is very spiky in nature. Also only the very last update usually matters, all previous updates can be often just ignored.

The client is very sensitive to message processing delays. I have already cythonized the critical path in business logic and that gave great results. The bottleneck now is within ClientWebSocketResponse + FlowControlDataQueue
I see a few performance issues there:

1. FlowControlDataQueue internals are private, there is no  way to see if more message have been already received and are just waiting in the queue. If there are 20 messages in the queue, I don't need all of them, only the last one.

2. FlowControlDataQueue, can potentially pause reading. There is no way to monitor, control or turn it off.

3. ClientWebSocketResponse.receive setup timeouts, creates and set result on a future even if there are more messages in FlowControlDataQueue. That's unnecessary.

4. I believe cythonizing WebSocketReader would help. parse_frame seems to be a great candidate for cythonization. I would also cythonize and include into distribution a pxd for WSMessage type and relevant enums

Ideally, would be great if I could pass my own WSMessage handler with a simple non-async on_ws_message callback and bypass the whole ClientWebSocketResponse + FlowControlDataQueue logic. Instead of optimizing ClientWebSocketResponse, I would rather have such a low level interface. 

Something similar to Protocol.data_received, Protocol.eof_received from asyncio. These are low-level non-async callbacks. Would be nice to have something similar for websockets.

**Comment by bdraco** [comment]
There have been some improvements in 3.10.x.  Some of the above is still quite relevant though.

**Reply by bdraco** [reply]
3.10.3 should improve things a bit more as well for some use cases

**Reply by tarasko** [reply]
> 3.10.3 should improve things a bit more as well for some use cases

Thank you. 
I ended up making my own small WebSockets library. 
https://github.com/tarasko/picows
Check out the benchmark I made, still a lot of room for improvement in aiohttp.

**Reply by bdraco** [reply]
Thanks for the benchmarks.  When I benchmark Home Assistant's use case I see that most of the time is spent in `http_websocket.py` with the top hits being `parse_frame`, followed by `_feed_data`

https://github.com/aio-libs/aiohttp/issues/2837 has some thoughts about making a cythonized `parse_frame`.  Since you have already done that work for `picows`, is that something you might be willing to contribute to `aiohttp`?

**Reply by tarasko** [reply]
Maybe, can't promise, if I find time for this. 

The issue with aiohttp (for me) is not just inefficient parse_frame but also FlowControlDataQueue and async iteration interface, which comes at cost. As I mentioned I'd rather have low level callback based access to parsed frames.

Also in my case it is possible to drop all messages except for the last one. Without knowing that I would have to parse the content of every message which is costly itself.

All in all I have quite a special use-case and having my own library was the easiest solution.

**Reply by bdraco** [reply]
No worries. It's no surprise that your purpose-built library is faster for your use case. :) I kind of expected that you wouldn't want to, given the time investment you had to put into that library, but I figured I'd ask just in case.

We have been discussing using `mypyc` to speed up the frame parsing in https://github.com/aio-libs/aiohttp/pull/8682#issuecomment-2284332460. Hopefully, that pans out.

Changing `FlowControlDataQueue` is a bit more challenging to do since we would end up exposing that as a public API, which we probably don't want to do at this point until we are very sure it is something we would like to live with for a long time. If we to a solution here, I'll update this discussion.

**Reply by tarasko** [reply]
Interesting, I haven't heard about mypyc.

Tbh, I'm fine with cython, I like that if something is difficult to implement in cython (due to lack of syntax for example) I can write plain c/c++ code in a separate file and just include it.

One thing that I'd like to look at in the near future is using SSE2/AVX2 instructions for masking frame payload. websockets library actually implemented it, at least for SSE2. 

Another one is uvloop. I have an impression that writing side could be a little more efficient than it is now in terms of latencies especially when sending small messages. But I haven't profiled it yet

**Reply by tarasko** [reply]
So regarding uvloop, I found that just removing this last condition 
`self._buffer_size > self._high_water`
here https://github.com/MagicStack/uvloop/blob/4083a94e5c1c3ab681ffe0b2c0b60fba16540220/uvloop/handles/stream.pyx#L426
increases websocket echo perfromance for aiohttp by 20-25% in my benchmark

So basically instead of caching and sending even small packets later, we send them right away. 
I will make a PR for uvloop

**Reply by bdraco** [reply]
I kinda surprised uvloop was doing that given https://github.com/python/cpython/blob/c13e7d98fb8581014a225b900b1b88ccbfc28097/Lib/asyncio/selector_events.py#L1065

**Reply by tarasko** [reply]
Yeah, asyncio implementation makes perfect sense to me, simple and concise. I'll try to simplify this part in uvloop

**Reply by tarasko** [reply]
https://github.com/MagicStack/uvloop/pull/619
